draft-ralston-mimi-linearized-matrix.md

```diff
@@ -687,17 +687,215 @@ The hub server MUST enforce the room version's algorithms upon the room. Partici
+* A device can be removed by anyone if the user ID to which it belongs is no longer in the `join` state.
+* A device cannot be removed any other way, as per MLS.
+
+If this validation fails, the hub server MUST reject the request if it's shaped as an LPDU ({{int-lpdu}})
+and soft-fail ({{int-soft-failure}}) the event if it's a PDU ({{int-pdu}}).
+
+Welcome messages are sent to devices over to-device messaging ({{int-transport-to-device}}). The `type`
+for the message is `m.room.encrypted` [**TODO**: Rename to avoid confusion with room event?] and `content`
+of:
+
+~~~ json
+{
+   "algorithm": "m.mls.v1.welcome.[ciphersuite]",
+   "ciphertext": "<unpadded base64 encoded welcome message>",
+   "commit_event_id": "<event ID of the m.mls.commit event>"
+}
+~~~
+
+`algorithm` is the ciphersuite, `dhkemx25519-aes128gcm-sha256-ed25519`, prefixed with `m.mls.v1.welcome`.
+
+The remaining fields are as described in the example. See {{int-unpadded-base64}} for "unpadded base64".
+
+All fields MUST be supplied. Note that the sender's user ID and device ID are made available over the
+to-device messaging endpoints ({{int-transport-to-device}}).
+
+In all cases, a device remembers the event ID (either from the `m.mls.commit` event or `commit_event_id`
+from a to-device message) after decryption to associate it with the MLS epoch. The device can then do
+a reverse lookup of epoch to event ID to MLS group state. Note that a client *always* has access to
+`m.mls.commit` events, even when hidden by history visibility ({{int-calc-event-visibility}}).
+
+**TODO**: We may need to store the group state in the media repo if it gets to be too big, or otherwise
+allow oversized events.
+
+## Key Packages {#int-mls-key-packages}
+
+Clients "claim" another device's key package through their server ({{int-transport-key-claim}}). Clients
+will typically generate several key packages and upload them to their server, making them available even
+if the client goes offline.
+
+The algorithm for a key package is `m.mls.v1.key_package.dhkemx25519-aes128gcm-sha256-ed25519` and is
+combined with a device-generated key version, forming a key ID described by {{int-signing}}. The key
+version SHOULD be generated based upon the key package itself rather than using an unrelated string,
+such as a hash or the public key of the key package.
+
+## Room Event Types
+
+This document describes the following event types for use with MLS-encrypted rooms. The section headers
+are the event `type`. See {{int-pdu}} for more information on events.
+
+These event types are non-state events, also called "room events".
+
+### `m.mls.commit` {#int-ev-mls-commit}
+
+Represents an MLS commit, which may be rejected by the hub server.
+
+`content` for the event MUST contain at least the following example:
+
+~~~ json
+{
+   "auth": "<unpadded base64 encoded FramedMessageAuth>",
+   "message": "<unpadded base64 encoded FramedMessage>"
+}
+~~~
+
+**TODO**: Verify shape of `content`.
```

Comment:
Reviewer: please verify

Reply:
I'm not sure what `FramedMessageAuth` and `FramedMessage` are, but I'm guessing they're parts of the MLS message?  We should probably just use the whole serialized MLS `PublicMessage` as a single chunk.

We also need a `commit_event_id` property, pointing to the event ID where the parent epoch was created.  This is to allow for backfilling.  i.e. the client gets a `limited` sync, and so can't decrypt the latest messages because it hasn't created those epochs yet.  So it looks up the `commit_event_id` to get the commit, and does so recursively until it gets to a commit that it can process. (or realizes that it wasn't in the group at the time, and so can't decrypt)

Reply:
Also we need a property for the public group state, if we need to support external joins.

Reply:
`FramedMessage*` comes from the MLS protocol spec, which makes up a `PublicMessage`. I'm sure there was a reason why I broke it down, but equally we don't seem to need to, so meh. 

Is the public room state not part of the `PublicMessage` parts? If we need it as an additional field, can add it.

I might call the commit ID field `prev_commit_event_id` though to denote it's pointing at a predecessor. We might also need to shorten some of these names eventually.

Reply:
I'm seeing `FramedContent` and `FramedContentAuthData` at https://github.com/mlswg/mls-protocol/blob/main/draft-ietf-mls-protocol.md#encoding-and-decoding-a-public-message and https://messaginglayersecurity.rocks/mls-protocol/draft-ietf-mls-protocol.html#name-encoding-and-decoding-a-pub

Part of the public room state is in the `PublicMessage` part.  The `RatchetTree` may be included, if the group uses the `RatchetTree` extension.  We may or may not want to use the `RatchetTree` extension.  One downside with it is that's the part of the public group state that can get really big, and could necessitate storing in the media repo.